### PR TITLE
feat(HPM): add basic HPMs for TL2TL and TL2CHI CoupledL2

### PR DESCRIPTION
**Pull request by yu-yake2002**
|  idx | Name  |
|  ----  | ----  |
| 1 | l2_cache_refill |
| 2 | l2_cache_rd_refill |
| 3 | l2_cache_wr_refill |
| 4 | l2_cache_long_miss |
| 5 | l2_cache_access |
| 6 | l2_cache_l2wb |
| 7 | l2_cache_l1wb |
| 8 | l2_cache_wb_victim |
| 9 | l2_cache_wb_cleaning_coh |
| 10 | l2_cache_access_rd |
| 11 | l2_cache_access_wr |
| 12 | l2_cache_inv |

### CODE REVIEW

File: src/main/scala/coupledL2/tl2chi/MSHRCtl.scala [hunk 1]
```diff
@@ -222,5 +223,30 @@ class MSHRCtl(implicit p: Parameters) extends TL2CHIL2Module with HasCHIOpcodes
...
+    val enable = m.io.status.valid && m.io.status.bits.will_free
+    when (enable && hpm_timer > 200.U) {
+      lmiss(i) := 1.U
+    }.otherwise {
+      lmiss(i) := 0.U
+    }
+  }
+
+  val perfEvents = Seq(
+    ("l2_cache_refill", io.resps.rxdat.valid && (io.resps.rxdat.respInfo.chiOpcode.get === CompData) && io.resps.rxdat.respInfo.last),
```
Comment (linjuanZ) [inline src/main/scala/coupledL2/tl2chi/MSHRCtl.scala:245]: Refill should also consider separate data and response. And what about dateless request?

File: src/main/scala/coupledL2/tl2chi/MainPipe.scala [hunk 2]
```diff
@@ -933,4 +934,16 @@ class MainPipe(implicit p: Parameters) extends TL2CHIL2Module with HasCHIOpcodes
 
   XSPerfAccumulate("early_prefetch", meta_s3.prefetch.getOrElse(false.B) && !meta_s3.accessed && !dirResult_s3.hit && task_s3.valid)
 
+  /* ===== Hardware Performance Monitor ===== */
+  val perfEvents = Seq(
+    ("l2_cache_access", task_s3.valid && (sinkA_req_s3 && !req_prefetch_s3 || sinkC_req_s3)),
+    ("l2_cache_wb", task_s3.valid && (mshr_cbWrData_s3 || mshr_snpRespDataX_s3)),
+    ("l2_cache_allocate", task_s3.valid && sinkC_req_s3 && (req_s3.opcode === ReleaseData)),
```
Comment (linjuanZ) [inline src/main/scala/coupledL2/tl2chi/MainPipe.scala:941]: What does `l2_cache_allocate` mean?

File: src/main/scala/coupledL2/utils/PerfCounterUtils.scala
Comment (linjuanZ) [inline src/main/scala/coupledL2/utils/PerfCounterUtils.scala:1]: Please move this (and also `PerfCounterUtils.scala` in XiangShan) into utility repo.

**Review by linjuanZ — CHANGES_REQUESTED** [review]

File: src/main/scala/coupledL2/tl2chi/MSHRCtl.scala [hunk 3]
```diff
@@ -222,5 +223,30 @@ class MSHRCtl(implicit p: Parameters) extends TL2CHIL2Module with HasCHIOpcodes
...
+
+  val hpm_timers = RegInit(VecInit(Seq.fill(mshrsAll)(0.U(10.W))))
+  // TODO: Is the width(4.W) proper here?
+  val lmiss = Wire(Vec(mshrsAll, UInt(4.W)))
+  for (((hpm_timer, m), i) <- hpm_timers.zip(mshrs).zipWithIndex) {
+    when (m.io.alloc.valid) {
+      hpm_timer := 1.U
+    }.otherwise {
+      hpm_timer := hpm_timer + 1.U
+    }
```
[the remark this answers is not in the document]
Comment (yu-yake2002) [reply src/main/scala/coupledL2/tl2chi/MSHRCtl.scala:235]: When an MSHR is allocated, line 232 will set `hpm_timer` to 1. It's similar to XSPerf named `mshr_latency` in tl2tl's MSHRCtl.

File: src/main/scala/coupledL2/tl2chi/MSHRCtl.scala [hunk 4]
```diff
@@ -222,5 +223,30 @@ class MSHRCtl(implicit p: Parameters) extends TL2CHIL2Module with HasCHIOpcodes
...
+    }.otherwise {
+      lmiss(i) := 0.U
+    }
+  }
+
+  val perfEvents = Seq(
+    ("l2_cache_refill", io.resps.rxdat.valid && (io.resps.rxdat.respInfo.chiOpcode.get === CompData) && io.resps.rxdat.respInfo.last),
+    ("l2_cache_rd_refill", io.resps.rxdat.valid && (io.resps.rxdat.respInfo.chiOpcode.get === CompData) && io.resps.rxdat.respInfo.last),
+    ("l2_cache_wr_refill", false.B),
+    ("l2_cache_lmiss", lmiss.reduce(_ + _))
```
[the remark this answers is not in the document]
Comment (yu-yake2002) [reply src/main/scala/coupledL2/tl2chi/MSHRCtl.scala:248]: Long miss

File: src/main/scala/coupledL2/tl2chi/MainPipe.scala (hunk 2 above)
Comment (yu-yake2002) [reply src/main/scala/coupledL2/tl2chi/MainPipe.scala:941]: Actually this name is from ARM's manual.😂
It means L2 data cache allocation without refill. I think this name is confusing, too. We can rename this event.

File: src/main/scala/coupledL2/tl2chi/MSHRCtl.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (yu-yake2002) [reply src/main/scala/coupledL2/tl2chi/MSHRCtl.scala:235]: I don't think it's necessary to clear the timer. After an MSHR is deallocated, its `lmiss` will be disabled, until it's allocated again.

File: src/main/scala/coupledL2/utils/PerfCounterUtils.scala
Comment (yu-yake2002) [reply src/main/scala/coupledL2/utils/PerfCounterUtils.scala:1]: In Utility: https://github.com/OpenXiangShan/Utility/pull/81
In Xiangshan: https://github.com/OpenXiangShan/XiangShan/pull/3631
In CoupledL2: https://github.com/OpenXiangShan/CoupledL2/pull/251/commits/cba7a2bcf36206af889455ca1943bb87fe5a6f5f

File: src/main/scala/coupledL2/tl2chi/MSHRCtl.scala (hunk 1 above)
Comment (yu-yake2002) [reply src/main/scala/coupledL2/tl2chi/MSHRCtl.scala]: Fixed in https://github.com/OpenXiangShan/CoupledL2/pull/251/commits/1d073e51d3c51aff2609ce1c6dbb9705bf6cb063. Dataless requests aren't included here.

**Review by linjuanZ — APPROVED** [review]

**Comment by yu-yake2002** [comment]
It seems that OpenLLC need to bump utility as well.🥲

**Review by Ivyfeather — APPROVED** [review]

**Pull request merged**